proof/refine/KHeap_R.thy

```diff
@@ -11,4 +11,2094 @@ imports
+           (loadObject_default p q n ko :: ('a::pre_storable) kernel)"
+  apply (simp add: loadObject_default_def split_def projectKO_def
+                   alignCheck_def alignError_def magnitudeCheck_def
+                   unless_def)
+  apply (rule no_fail_pre)
+   apply (wp case_option_wp)
+  apply (clarsimp simp: is_aligned_mask)
+  apply (clarsimp split: option.split_asm)
+  apply (clarsimp simp: is_aligned_mask[symmetric])
+  apply simp
+  done
+
+lemma no_fail_getObject_tcb[wp]:
+  "no_fail (tcb_at' t) (getObject t :: tcb kernel)"
+  apply (simp add: getObject_def split_def)
+  apply wp
+  apply (clarsimp simp: obj_at'_def gen_objBits_simps)
+  apply (frule is_aligned_no_overflow)
+  apply (rule ps_clear_lookupAround2, assumption+)
+    apply simp+
+  apply (fastforce split: option.split_asm simp: gen_objBits_simps)
+  done
+
+lemma corres_get_tcb:
+  "corres (tcb_relation \<circ> the) (tcb_at t) (tcb_at' t) (gets (get_tcb t)) (getObject t)"
+  apply (rule corres_no_failI)
+   apply wp
+  apply (clarsimp simp add: gets_def get_def return_def bind_def get_tcb_def)
+  apply (frule in_inv_by_hoareD [OF getObject_inv_tcb])
+  apply (clarsimp simp add: obj_at_def is_tcb obj_at'_def projectKO_def
+                            projectKO_opt_tcb split_def
+                            getObject_def loadObject_default_def in_monad)
+  apply (case_tac koa)
+   apply (simp_all add: fail_def return_def)
+  apply (case_tac bb)
+   apply (simp_all add: fail_def return_def)
+  apply (clarsimp simp add: state_relation_def pspace_relation_def)
+  apply (drule bspec)
+   apply clarsimp
+   apply blast
+  apply (clarsimp simp: tcb_relation_cut_def lookupAround2_known1)
+  done
+
+lemma lookupAround2_same1[simp]:
+  "(fst (lookupAround2 x s) = Some (x, y)) = (s x = Some y)"
+  apply (rule iffI)
+   apply (simp add: lookupAround2_char1)
+  apply (simp add: lookupAround2_known1)
+  done
+
+lemma getObject_tcb_at':
+  "\<lbrace> \<top> \<rbrace> getObject t \<lbrace>\<lambda>r::tcb. tcb_at' t\<rbrace>"
+  by (fastforce simp: valid_def getObject_def in_monad
+                      loadObject_default_def obj_at'_def split_def
+                      in_magnitude_check gen_objBits_simps
+                split: kernel_object.splits)
+
+lemma updateObject_cte_is_tcb_or_cte:
+  fixes cte :: cte and ptr :: machine_word
+  shows "\<lbrakk> fst (lookupAround2 p (ksPSpace s)) = Some (q, ko);
+           snd (lookupAround2 p (ksPSpace s)) = n;
+           (ko', s') \<in> fst (updateObject cte ko p q n s) \<rbrakk> \<Longrightarrow>
+  (\<exists>tcb getF setF. ko = KOTCB tcb \<and> s' = s \<and> tcb_cte_cases (p - q) = Some (getF, setF)
+    \<and> ko' = KOTCB (setF (\<lambda>x. cte) tcb) \<and> is_aligned q tcbBlockSizeBits \<and> ps_clear q tcbBlockSizeBits s) \<or>
+  (\<exists>cte'. ko = KOCTE cte' \<and> ko' = KOCTE cte \<and> s' = s
+        \<and> p = q \<and> is_aligned p cte_level_bits \<and> ps_clear p cte_level_bits s)"
```

Comment:
More style/indentation that it would be nice to fix while we're doing this.

Reply:
done